### ISSUE TITLE
Unstable type inference when generic is inferred from a return type

### ISSUE THREAD

**Issue by Jessidhia** [Microsoft/TypeScript#29771]
**TypeScript Version:**  3.4.0-dev.20190206

**Search Terms:** generic return

**Code**

```ts
declare function useDerivedState<T extends unknown>(
  initializer: (previous: T | undefined) => T,
  deps?: ReadonlyArray<unknown>
): T

function makeFoo(previous: { foo?: string } = {}) {
  return {
    ...previous,
    foo: 'bar'
  }
}

const state = useDerivedState(previous => makeFoo(previous))
state.foo
```

**Expected behavior:**

`foo` exists and is accessible

**Actual behavior:**

Very inconsistent.

This image is probably the most egregious example (`state` is `{ foo: string }`):

<img width="470" alt="image" src="https://user-images.githubusercontent.com/73085/52351087-59ecbe00-2a6d-11e9-9567-2f3bc48c98a3.png">

But if I then type a whitespace or something to get the compiler to update, sometimes it starts seeing `state` as `{}`:

<img width="155" alt="image" src="https://user-images.githubusercontent.com/73085/52351201-94eef180-2a6d-11e9-9308-7142ce385204.png">

There are also some cases I can't screenshot properly. For example, this is the full derivation when it is almost working correctly:

<img width="346" alt="image" src="https://user-images.githubusercontent.com/73085/52351357-f2833e00-2a6d-11e9-990e-7a3c08e583b7.png">

But if I cause some input to happen, the derivation might reset to `<{}>`. However, pressing command to be able to use the screenshot tool causes the tooltip to update and it goes back to the full derivation.

The only thing that is consistent is that type of the `previous` argument seems to always be derived as `{}`.

**Related Issues:** https://github.com/Microsoft/TypeScript/issues/29638 and probably a bunch of others that look similar but don't quite show this odd / pathological behavior.

**Comment by dragomirtitian** [comment]
A samller example of what you are trying to do would be:

```ts

declare function useDerivedState<T>(
  initializer: (previous: T ) => T
): T

const state = useDerivedState(previous => 0) // state is {}
```

Just as an opinion, I don't think typescript can deal with this scenario well at the moment. It's probably trying to figure out the type of the callback, and to do that it needs a type for the parameter, and there is nowhere to get that from so goes with the default of `{}`.

**Comment by Jessidhia** [comment]
Yeah, that is probably the central cause, but the odd thing is that the inference was so unstable and would change by just interacting with the IDE even without changing the source code.

**Comment by RyanCavanaugh** [comment]
Our inference system can't really deal with a `t => t'` kind of function to do inferences from both sides to figure out what's going on. Sometimes language service operations manage to avoid noticing, but this is really a side effect.

**Comment by Weakky** [comment]
Same here, which is very sad 😞

```ts
interface ServerConfig<T, U, V> {
  server: () => T
  startServer: (param: T) => U
  stopServer: (param: U) => V
}

function config<T, U, V>(cfg: ServerConfig<T, U, V>): ServerConfig<T, U, V> {
  return cfg
}

const serverConfig = config({
    server: () => {
        return { foo: 'bar' }
    },
    startServer: (param) => {
        return param.foo
    },
    stopServer: (param) => {
        return 5
    }
})
```

![image](https://user-images.githubusercontent.com/25233379/54116924-1086ea00-43f0-11e9-98a7-9fddca8b3a9c.png)

**Expected behavior:**

`ServerConfig` to be of type

```ts
ServerConfig<{ foo:'bar' }, string, number>
```

**Actual behavior:**

```ts
ServerConfig<{ foo:'bar' }, {}, number>
```

[Playgroung link](http://www.typescriptlang.org/play/#src=type%20Omit%3CT%2C%20K%20extends%20keyof%20T%3E%20%3D%20Pick%3CT%2C%20Exclude%3Ckeyof%20T%2C%20K%3E%3E%3B%0A%0Atype%20Func%3CT%3E%20%3D%20(arg%3A%20T)%20%3D%3E%20null%0A%0Atype%20Context%20%3D%20'Context'%3B%0A%0Aexport%20function%20withRouteContextPropConsumer%3C%0A%20%20%20%20T%20extends%20%7B%20routeContext%3A%20Context%20%7D%0A%3E(%0A%20%20%20%20funcToWrap%3A%20Func%3CT%3E%2C%0A)%3A%20Func%3COmit%3CT%2C%20%22routeContext%22%3E%3E%20%7B%0A%20%20%20%20return%20(args%3A%20Omit%3CT%2C%20%22routeContext%22%3E)%20%3D%3E%20%7B%0A%20%20%20%20%20%20%20%20const%20routeContext%3A%20Context%20%3D%20'Context'%3B%0A%20%20%20%20%20%20%20%20return%20funcToWrap(%7B%20...args%2C%20routeContext%20%7D)%3B%0A%20%20%20%20%7D%3B%0A%7D)

The return types are sometimes inferred, sometimes not, which makes it unreliable 😢

**Comment by MicahZoltu** [comment]
Oddly, when you give the compiler _less_ information it gets the inference correct.  I don't know why this is, and perhaps is related to the "unreliability" that others were talking about:
```typescript
interface Apple<T> {
  parse: (value: string) => T,
  stringify: (value: T) => string,
};
declare function eatApple<T>(apple: Apple<T>): void
eatApple({
    // infers `T` as `unknown`
    parse: value => Number.parseInt(value),
    // errors because `value` is `unknown`
    stringify: value => value.toString(),
})
eatApple({
    // infers `T` as `number`
    parse: (value: string) => Number.parseInt(value),
    // works because `value` is `number`
    stringify: value => value.toString(),
})

interface Banana<T> {
    parse: (value: string) => T
}
declare function eatBanana<T>(banana: Banana<T>): void
eatBanana({
    // `T` correctly inferred as `number`
    parse: value => Number.parseInt(value),
})
```
Of note is that an instance of `Apple` incorrectly inferred `T` as `unknown` when given an untyped `parse` function, however `Banana` correctly inferred `T` as `number` when given _the exact same function_.  The only difference between the two is that `Apple` has additional properties.

**Comment by threehams** [comment]
That makes sense, though. Inference only breaks down when the generic is used in both a return type position and another position.
